process/block/preprocess/selectionSession.go

```diff
@@ -38,67 +39,80 @@ func NewSelectionSession(args ArgsSelectionSession) (*selectionSession, error) {
 	return &selectionSession{
 		accountsAdapter:        args.AccountsAdapter,
 		transactionsProcessor:  args.TransactionsProcessor,
-		ephemeralAccountsCache: make(map[string]vmcommon.AccountHandler),
+		ephemeralAccountsCache: make(map[string]state.UserAccountHandler),
 	}, nil
 }
 
-// GetAccountState returns the state of an account.
-// Will be called by mempool during transaction selection.
-func (session *selectionSession) GetAccountState(address []byte) (state.UserAccountHandler, error) {
-	account, err := session.getExistingAccount(address)
+// GetAccountNonceAndBalance returns nonce and balance of an account.
+// Will be called by the transactions pool, during transactions selection.
+func (session *selectionSession) GetAccountNonceAndBalance(address []byte) (uint64, *big.Int, bool, error) {
+	account, err := session.getCachedUserAccount(address)
 	if err != nil {
-		// TODO: brainstorm whether we should move the logic around "ErrAccNotFound",
-		// from "virtualSelectionSession.createAccountRecord()", to this place.
-		// If we do so, we hide (encapsulate) that information (error) within this component:
-		// good on one hand, but, on the other hand, that piece of information (error) might be needed in the "TxCache" at some point
-		// (to take some specific actions).
-		return nil, err
+		// Unexpected failure.
+		return 0, nil, false, err
 	}
-
-	userAccount, ok := account.(state.UserAccountHandler)
-	if !ok {
-		return nil, process.ErrWrongTypeAssertion
+	if check.IfNil(account) {
+		// New (unknown) account.
+		return 0, big.NewInt(0), false, nil
 	}
 
-	return userAccount, nil
+	return account.GetNonce(), account.GetBalance(), true, nil
 }
 
-func (session *selectionSession) getExistingAccount(address []byte) (vmcommon.AccountHandler, error) {
+func (session *selectionSession) getCachedUserAccount(address []byte) (state.UserAccountHandler, error) {
 	account, ok := session.ephemeralAccountsCache[string(address)]
 	if ok {
+		// Existing or new (unknown) account, previously-cached.
 		return account, nil
 	}
 
-	account, err := session.accountsAdapter.GetExistingAccount(address)
-	if err != nil {
+	account, err := session.getExistingAccountTypedAsUserAccount(address)
+	if err != nil && err != state.ErrAccNotFound {
+		// Unexpected failure (error different from "ErrAccNotFound").
+		// Account won't be cached.
 		return nil, err
 	}
 
+	// Existing account or new (unknown), we'll cache it (actual object or nil).
 	session.ephemeralAccountsCache[string(address)] = account
+
+	// Generally speaking, this isn't a good pattern: returning both nil (for unknown accounts), and a nil error.
```

Comment:
Not a great pattern, but this slightly peculiar decision is ~well encapsulated within `selectionSession`.